### PR TITLE
Add Interpolation Expression Sequences to the spec

### PR DESCRIPTION
**Pull request by schveiguy**
See dlang/dmd#15715

### CODE REVIEW

**Comment by schveiguy** [comment]
OK, I think this is ready for review. This needs to be merged before the next release.

"Interpolation Expression Sequence" is too long for the menu bar... Any ideas for an abbreviated name? Maybe just "Interpolations"?

**Comment by schveiguy** [comment]
Unsure if adding this in where lexical goes (which I think is where it really belongs) is better than tacking on the end? Walter suggested [here](https://github.com/dlang/dmd/pull/15715#pullrequestreview-1824895012) that the spec should be on its own page.

I dislike that I had to bump all chapters up by 1.

File: spec/istring.dd [hunk 1]
```diff
@@ -0,0 +1,228 @@
+Ddoc
+
+$(SPEC_S Interpolation Expression Sequence,
+
+$(HEADERNAV_TOC)
+
+$(P Interpolation Expression Sequences (IES) are expressions that intersperse
+string literal data, and values. An interpolation expression is written like a
```
Comment (adamdruppe) [inline spec/istring.dd:8]: i'd delete the comma lol

otherwise tl;dr but no gross errors jumped out at me

**Comment by schveiguy** [comment]
ping, this should go in before the release candidate is branched.

**Review by thewilsonator — APPROVED** [review]

File: spec/istring.dd [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+
+------
+// simple version of std.typecons.Tuple
+struct Tuple(T...) { T value; }
+Tuple!T tuple(T...)(T value) { return Tuple!T(value); }
+
+import core.interpolation;
+string name = "John Doe";
+auto items = tuple(i"Hello, $(name), how are you?");
+assert(items == tuple(
```
Comment (ntrel) [inline spec/istring.dd:127]: Pretty sure the above lines can be simplified to:
```d
alias AliasSeq(A...) = A;
import core.interpolation;
auto items = i"Hello, $(name), how are you?";
assert(items == AliasSeq!(
```

Comment (schveiguy) [reply spec/istring.dd:127]: I will try it.

Comment (schveiguy) [reply spec/istring.dd:127]: Yes, this works, I will simplify.

Comment (schveiguy) [reply spec/istring.dd:127]: Hm... on second thought, this is doing something that I wasn't intending -- it's creating a compile-time sequence of things, including an alias to `name`. What I really wanted was to show that the sequence is the same at runtime, e.g. what the list of values is. So I think I will leave it as-is.

File: spec/istring.dd [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+escapes are recognized inside these literals.)
+
+$(H3 $(LNAME2 ies_token, Token IES Literals))
+
+$(GRAMMAR
+$(GNAME InterpolatedTokenLiteral):
+    $(D iq{) $(GLINK InterpolatedTokenStringTokens)$(OPT) $(D })
+
+$(GNAME InterpolatedTokenStringTokens):
+    $(GLINK InterpolatedTokenStringTokenNoBraces)
```
Comment (tim-dlang) [inline spec/istring.dd:88]: InterpolatedTokenStringTokenNoBraces is not defined. Here is a PR: https://github.com/dlang/dlang.org/pull/3772

**Pull request merged**